src/Grids/vertically_stretched_rectilinear_grid.jl

```diff
@@ -272,8 +272,8 @@ short_show(grid::VerticallyStretchedRectilinearGrid{FT, TX, TY, TZ}) where {FT,
     "VerticallyStretchedRectilinearGrid{$FT, $TX, $TY, $TZ}(Nx=$(grid.Nx), Ny=$(grid.Ny), Nz=$(grid.Nz))"
 
 function show(io::IO, g::VerticallyStretchedRectilinearGrid{FT, TX, TY, TZ}) where {FT, TX, TY, TZ}
-    Δz_min = minimum(view(g.Δzᵃᵃᶜ, 1:g.Nz))
-    Δz_max = maximum(view(g.Δzᵃᵃᶜ, 1:g.Nz))
+    Δz_min = minimum(view(parent(g.Δzᵃᵃᶜ), g.Hz+1:g.Nz+g.Hz))
+    Δz_max = maximum(view(parent(g.Δzᵃᵃᶜ), g.Hz+1:g.Nz+g.Hz))
```

Comment:
Could you elaborate on why this should eliminate the scalar operation error on GPUs?
(I'm mostly interested to learn... I'm not arguing that it's not the right way forward.)

Reply:
`g.Δzᵃᵃᶜ` is an `OffsetArray`, which is a wrapper around a `CuArray` that changes indexing behavior. (`parent(g.Δzᵃᵃᶜ) returns the underlying `CuArray`.) `view` also returns a wrapper that both shifts indices and defines new `size` and `length` functions.

So `view(g.Δzᵃᵃᶜ)` is a doubly nested wrapper.

For computing reductions on the GPU like `maximum` and `minimum`, `GPUArrays.jl` and `CUDA.jl` provide special implementations of `Base.mapreducedim!`, which performs in-place reductions of array-like objects. Then some common reduction functions are defined using this implementation

https://github.com/JuliaGPU/GPUArrays.jl/blob/7f8410a1147bfb25a552d8ab1ee85d74bf9ed6fa/src/host/mapreduce.jl#L83-L91

(this mirrors the implementation in `Base`.)

However, in those lines you'll notice that this special implementation is only provided for objects of type `AnyGPUArray`, defined here:

https://github.com/JuliaGPU/GPUArrays.jl/blob/4de9fbb3a64546f13834214f97b5ecfb4d94f1d3/src/host/abstractarray.jl#L23

This includes `WrappedGPUArray`, which is a large type union defined in `Adapt.jl` that includes many common "nested" array types, including `view`s into `CuArray`. But it only includes _singly_ nested arrays. It does not include _doubly_ nested arrays. The number of types that'd be needed in the type union to support _doubly_ nested arrays might be very long...

Because doubly nested arrays cannot use the special GPU-specified reduction functions (because they are not included in the type union `AnyGPUArray`), when we call `maximum` on a `view` into an `OffsetArray`, we fallback to the implementation in `Base`. And this implementation relies on indexing into the array, rather than launching a kernel that's GPU-friendly.

Reply:
See

https://github.com/JuliaGPU/Adapt.jl/blob/cac58aa57367cb6d607328d4679e7c2d39ddba61/src/wrappers.jl#L63

for the definition of `WrappedArray` in `Adapt`.

Bottom line is that in order to use the special GPU reductions we need to pass either raw `CuArray` or _singly_ wrapped `CuArray`, like `view(a::CuArray)` or `OffsetArray(a::CuArray)`. Doubly nested wrappers like `view(OffsetArray(a::CuArray))` cannot be used (at least not right not).

Reply:
The fact that `WrappedArray` only works with one level of nesting is mentioned here:

https://github.com/JuliaGPU/Adapt.jl/blob/cac58aa57367cb6d607328d4679e7c2d39ddba61/src/wrappers.jl#L101-L102

and can also be seen from the definition of `WrappedArray`.

Reply:
Also note that the "allocation" versions of reduction functions (eg `maximum`, compared with the in-place `maximum!`) work merely by first calling `similar(a)` (to allocate output) and then calling the in-place versions on that object). For the case of a scalar (`maximum(a, dims=:)`) we get the somewhat silly pattern

https://github.com/JuliaGPU/GPUArrays.jl/blob/7f8410a1147bfb25a552d8ab1ee85d74bf9ed6fa/src/host/mapreduce.jl#L64-L68

So in order to understand how all the reduction functions work at the outermost level, we should first focus on the in-place versions.